provisioning/baremetal_secrets.go

```diff
@@ -0,0 +1,131 @@
+	ironicUsernameKey   = "username"
+	ironicPasswordKey   = "password"
+	ironicHtpasswdKey   = "htpasswd"
+	ironicConfigKey     = "auth-config"
+	ironicSecretName    = "metal3-ironic-password"
+	ironicUsername      = "ironic-user"
+	inspectorSecretName = "metal3-ironic-inspector-password"
+	inspectorUsername   = "inspector-user"
+)
+
+func generateRandomPassword() (string, error) {
+	chars := []rune("ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+		"abcdefghijklmnopqrstuvwxyz" +
+		"0123456789")
+	length := 16
+	buf := make([]rune, length)
+	numChars := big.NewInt(int64(len(chars)))
+	for i := range buf {
+		c, err := rand.Int(rand.Reader, numChars)
+		if err != nil {
+			return "", err
+		}
+		buf[i] = chars[c.Uint64()]
+	}
+	return string(buf), nil
+}
+
+func createMariadbPasswordSecret(client coreclientv1.SecretsGetter, targetNamespace string) error {
+	_, err := client.Secrets(targetNamespace).Get(context.Background(), baremetalSecretName, metav1.GetOptions{})
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Secret does not already exist. So, create one.
+	password, err := generateRandomPassword()
+	if err != nil {
+		return err
+	}
+	_, err = client.Secrets(targetNamespace).Create(
+		context.Background(),
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      baremetalSecretName,
+				Namespace: targetNamespace,
+			},
+			StringData: map[string]string{
+				baremetalSecretKey: password,
+			},
+		},
+		metav1.CreateOptions{},
+	)
+	return err
+}
+
+func createIronicPasswordSecret(client coreclientv1.SecretsGetter, targetNamespace string, name string, username string, configSection string) error {
+	_, err := client.Secrets(targetNamespace).Get(context.Background(), name, metav1.GetOptions{})
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Secret does not already exist. So, create one.
+	password, err := generateRandomPassword()
+	if err != nil {
+		return err
+	}
+	hash, err := bcrypt.GenerateFromPassword([]byte(password), 5) // Use same cost as htpasswd default
+	if err != nil {
+		return err
+	}
+	// Change hash version from $2a$ to $2y$, as generated by htpasswd.
+	// These are equivalent for our purposes.
+	hash[2] = 'y'
+
+	_, err = client.Secrets(targetNamespace).Create(
+		context.Background(),
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: targetNamespace,
+			},
+			StringData: map[string]string{
+				ironicUsernameKey: username,
+				ironicPasswordKey: password,
+				ironicHtpasswdKey: fmt.Sprintf("%s:%s", username, hash),
+				ironicConfigKey: fmt.Sprintf(`[%s]
+auth_type = http_basic
+username = %s
+password = %s
+`,
+					configSection, username, password),
+			},
+		},
+		metav1.CreateOptions{},
+	)
+	log.V(1).Info("created password for", "name", name)
+	return err
+}
+
+// CreateMetal3PasswordSecrets creates all Secrets that a metal3 pod would need
+func CreateMetal3PasswordSecrets(client coreclientv1.SecretsGetter, targetNamespace string) error {
```

Comment:
Making this function the public API for the package is going to make it more difficult to change the behavior later if we change the metal3 deployment and don't need some of the secrets. I think I would just expose each of the other functions and eliminate this one. That will also make it easier to implement the reconcile logic for re-creating a secret if one is removed.

Should we also separate the step of creating the `Secret` in memory and the step of passing that to the client, since for unit tests we may want to verify that a `Secret` has particular members?

Reply:
Couple of issues listed in the comment above -
1. Support creating a subset of the secrets due to changes to CBO
2. Create only the secret that is not present during the reconcile loop
Suggestion is that the solution to both these problems is that we should get rid of the top level function and be able to call individual methods separately. 

Having the top level functions abstracts away the many constants in play while generating the password. This should not be within the reconcile logic. Also, these functions already check if the Secret is already there and will create a new one **only** if not present. So, that is a non-issue.

Last part of code is asking for code re-organization to help with unit testing. The unit tests can be augmented to check for more fields within the Secret without creating the Secret in memory 1st.

Reply:
Hopefully, addressed this in the latest changes.

Reply:
We should be restarting a Pod(s) when the secrets change. 
Also, there is currently no need for a method to delete a secret that was created. But, the current way the code is organized makes it easy to add those methods when we need them.
At this point we do not need to pass provisioning config to these methods because currently all secrets are needed for all ProvisioningNetwork modes. When that changes, we can always come back and update these methods.